### PR TITLE
Create Vagrant test framework

### LINKED ISSUE #1471: Create vagrant-test binary
Create a rust integration test binary that can interface with vagrant.

At a minimum it should be able to run the following commands:

`vagrant up [hosts]...`

`vagrant halt [hosts]...`

`vagrant snapshot save [host] <name>`

`vagrant snapshot delete [host] <name>`

`vagrant snapshot restore [host] <name>`

`vagrant provision --provision-with=<name> [hosts]...`

### LINKED ISSUE #1472: Create test setup / teardown fns
Abstract setup / teardown fns

- `setup_bare()`
    - `vagrant up iscsi adm mds1 mds2 oss1 oss2 c1`
    - `vagrant provision --provision-with=yum-update`
    - `vagrant halt`
    - `vagrant snapshot save {iscsi,adm,mds1,mds2,oss1,oss2} bare`
    - `vagrant up iscsi adm mds1 mds2 oss1 oss2 c1`

- `setup_iml_install()`
   - run `setup_bare_cluster` within this fn
   - `vagrant provision --provision-with=install-iml-local`
   - `vagrant halt`
   - `vagrant snapshot save {iscsi,adm,mds1,mds2,oss1,oss2} iml-installed`

- `setup_deploy_servers(hostlist:String, profile:String)`
    - run `setup_iml_install` within this fn
    - `ssh adm iml server add -h {hostlist} -p {profile}`
    - `vagrant snapshot save {iscsi,adm,mds1,mds2,oss1,oss2} iml-deployed`

### PR DESCRIPTION
**Pull request by jgrund**
This patch implements a rudimentary vagrant test framework.

Fixes #1471.
Fixes #1472.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1495)

### CODE REVIEW

**Comment by mkpankov** [comment]
Will play with it a little tomorrow

**Comment by jgrund** [comment]
To run with std{out,err} in terminal:

`cargo test -p iml-integration-tests -- --nocapture`

**Comment by mkpankov** [comment]
Tried it on Windows, got couple issues. Not 100% we should try solving them, but maybe they aren't Windows-specific. [First](https://gist.github.com/mkpankov/504284dd2e206b36cfd50600f26ae8d3). [second](https://gist.github.com/mkpankov/e7cf046bab8fea301b2c9d9175a305f9).

**Comment by mkpankov** [comment]
On Linux, got an [issue](https://gist.github.com/mkpankov/c16377b7d66de656ca2889ef6de177ab) that looks similar to second one on Windows.

File: .github/workflows/rust-ci.yml [hunk 1]
```diff
@@ -77,6 +77,7 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: test
+          args: --exclude iml-integration-tests --workspace
```
Comment (ip1981) [inline .github/workflows/rust-ci.yml]: Locally this might be inconvenient.

File: iml-integration-tests/tests/integration_test.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+mod vagrant;
+
+#[tokio::test]
+async fn test_setup() -> Result<(), Box<dyn std::error::Error>> {
+    let config = vagrant::ClusterConfig::default();
+
+    vagrant::setup_deploy_servers(&config, "base_managed_patchless").await?;
```
Comment (ip1981) [inline iml-integration-tests/tests/integration_test.rs]: Why a new wrapper binary in Rust instead of a simple shell script in `./vagrant`? Windows?

Comment (jgrund) [reply iml-integration-tests/tests/integration_test.rs]: This allows more flexibility in that the specific servers / profile can be specified programmatically.

File: .github/workflows/rust-ci.yml (hunk 1 above)
Comment (jgrund) [reply .github/workflows/rust-ci.yml]: Open to other suggestions.

Comment (ip1981) [reply .github/workflows/rust-ci.yml]: Maybe it shouldn't be implemented as a test, but a standalone executable?

Comment (ip1981) [reply .github/workflows/rust-ci.yml]: There is also `#[ignored]` and `cargo test -- --ignored`.

**Comment by jgrund** [comment]
___
*[.github/workflows/rust-ci.yml, line 80 at r1](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1495#-Lz2dQt5n6lr0XIcIgmo-r1-80:-LzHgd513W-yPsxbMbVW:bm79imz) ([raw file](https://github.com/whamcloud/integrated-manager-for-lustre/blob/450d2122f263a3dcbf5a4ce6c8660f6efd960674/.github/workflows/rust-ci.yml#L80)):*
I thought about executable initially as well, however I think using `cargo test` will ultimately be better. 

Right now this only has one test, but it will eventually have many tests, at which point having something like cargo test to drive it will be better than a standalone where we write our own test runner.

In addition, we'll want to hook into test output and display test results in Jenkins which again wouldn't really make sense without a test runner imo.

> We don't gain much from hooking into cargo test, and this feels like repurposing.

I'm not sure I understand this, isn't the point of `cargo test` to run tests?


> Our tests aren't actually integration, they're system ones.

Sure, I'll rename the module.

File: .github/workflows/rust-ci.yml (hunk 1 above)
Comment (ip1981) [reply .github/workflows/rust-ci.yml]: My main concern is the possibility to step on the rake accidentally, in my view having these thing in a single repository and within single cargo workspace and use of the same testing framework make it more likely, but other than this I have no objections. I realized that I don't run tests ~~so I don't care~~ in a batch anyway (only per crate I am working on), thus it is ok :)

**Comment by nlinker** [comment]
___
*[.github/workflows/rust-ci.yml, line 80 at r1](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1495#-Lz2dQt5n6lr0XIcIgmo-r1-80:-Lzg4q6RECTjwI-MO2q7:b-di8n4a) ([raw file](https://github.com/whamcloud/integrated-manager-for-lustre/blob/450d2122f263a3dcbf5a4ce6c8660f6efd960674/.github/workflows/rust-ci.yml#L80)):*
<details><summary><i>Previously, ip1981 (Igor Pashev) wrote…</i></summary><blockquote>

My main concern is the possibility to step on a rake accidentally, in my view having these thing in a single repository and within single cargo workspace and use of the same testing framework make it more likely, but other than this I have no objections. I realized that I don't run tests ~~so I don't care~~ in a batch anyway (only per crate I am working on), thus it is ok :)
</blockquote></details>

My position is close to Joe's, `cargo test` reusage might give us many benefits, e.g. CI support.
An additional variant we might consider is the following:
```
cargo test   # runs the ordinary test suite

cargo test --system-tests   # runs the new tests against locally deployed vagrant vms
```

**Comment by jgrund** [comment]
___
*[.github/workflows/rust-ci.yml, line 80 at r1](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1495#-Lz2dQt5n6lr0XIcIgmo-r1-80:-M2P3MXl7DlEpaoGirGq:bcyymyo) ([raw file](https://github.com/whamcloud/integrated-manager-for-lustre/blob/450d2122f263a3dcbf5a4ce6c8660f6efd960674/.github/workflows/rust-ci.yml#L80)):*
<details><summary><i>Previously, mkpankov (Michael Pankov) wrote…</i></summary><blockquote>

```
cargo test --system-tests
```

That's not possible, I think? `cargo test` has its own flags we don't hook into. We can have `cargo test -- --system-tests` to pass our flags to the test binary, however, these flags are actually for `cargo` itself if they are to work.

Looking at its flags, what we can afford is probably something like
```
cargo test --test system_test_1
```

And now this test should be always run manually separately from other tests, because most of them can be run in parallel, but those with vagrant can't.

In this option, I don't like having to know the test's name.

We can probably also have

```
cargo test system_test_ -- --test-threads 1
```
to run all tests with names containing `system_test_` one after another.

I guess I'm not satisfied with default cargo test runner to begin with, but if we want to spend zero time on developing a test harness, we should make something from above options work.
</blockquote></details>

I've removed iml-system-tests from the workspace. It will not run by default.

I do agree that we should have a custom harness, but right now we have no automated testing for vagrant, and this will be *something* and a step in the right direction. After this lands, we can think about how to write a proper harness.

File: iml-system-tests/tests/integration_test.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) 2019 DDN. All rights reserved.
```
Comment (johnsonw) [inline iml-system-tests/tests/integration_test.rs]: Copyright year

File: iml-system-tests/tests/vagrant.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2019 DDN. All rights reserved.
```
Comment (johnsonw) [inline iml-system-tests/tests/vagrant.rs]: Year

File: iml-system-tests/tests/integration_test.rs (hunk 3 above)
Comment (jgrund) [reply iml-system-tests/tests/integration_test.rs]: We'll need to pass through and update everything in one patch.

**Pull request merged**